SentinelOne/sentinelone_module/logs/connector.py

```diff
@@ -0,0 +1,301 @@
+
+        self.connector._executor.shutdown(wait=True)
+
+
+class SentinelOneActivityLogsConsumer(SentinelOneLogsConsumer):
+    def pull_events(self):
+        """Fetches activities from SentinelOne"""
+        # Set  filters
+        query_filter = ActivitiesFilter()
+        query_filter.apply(key="limit", val=1000)
+        query_filter.apply(key="sortBy", val="createdAt")
+        query_filter.apply(key="sortOrder", val="asc")
+
+        while self.running:
+            # Fetch activities
+            activities = self.management_client.activities.get(query_filter)
+            INCOMING_MESSAGES.labels(intake_key=self.configuration.intake_key, datasource="sentinelone").inc(
+                len(activities.data)
+            )
+
+            # Push events
+            self.connector.push_events_to_intakes(self._serialize_events(activities.data))
+
+            # Update context with latest event date
+            latest_event_timestamp = self._get_latest_event_timestamp(activities.data)
+            with self.context as cache:
+                cache["last_event_date"] = latest_event_timestamp.isoformat()
+
+            # Send Prometheus metrics
+            OUTCOMING_EVENTS.labels(intake_key=self.configuration.intake_key, datasource="sentinelone").inc(
+                len(activities.data)
+            )
+            EVENTS_LAG.labels(intake_key=self.configuration.intake_key, type="activities").set(
+                (datetime.now(UTC) - latest_event_timestamp).total_seconds()
+            )
+
+            if activities.pagination["nextCursor"] is None:
+                break
+
+            query_filter.apply(key="cursor", val=activities.pagination["nextCursor"])
+
+
+class SentinelOneThreatLogsConsumer(SentinelOneLogsConsumer):
+    def pull_events(self):
+        """Fetches threats from SentinelOne"""
+        query_filter = ThreatQueryFilter()
+        query_filter.apply(key="limit", val=1000)
+        query_filter.apply(key="sortBy", val="createdAt")
+        query_filter.apply(key="sortOrder", val="asc")
+
+        while self.running:
+            # Fetch threats
+            threats = self.management_client.client.get(endpoint="threats", params=query_filter.filters)
+
+            # Push events
+            self.connector.push_events_to_intakes(self._serialize_events(threats.data))
+
+            # Update context with the latest event date
+            latest_event_timestamp = self._get_latest_event_timestamp(threats.data)
+            with self.context as cache:
+                cache["last_event_date"] = latest_event_timestamp.isoformat()
+
+            # Send Prometheus metrics
+            OUTCOMING_EVENTS.labels(intake_key=self.configuration.intake_key, datasource="sentinelone").inc(
+                len(threats.data)
+            )
+            EVENTS_LAG.labels(intake_key=self.configuration.intake_key, type="threats").set(
+                (datetime.now(UTC) - latest_event_timestamp).total_seconds()
+            )
```

Comment:
If no events are fetched, the method `_get_latest_event_timestamp` will return the timestamp stored in the context. So, if no events are fetched, the events lag will increase until the next collected events.
Can you change the way to compute the events lag if no events are fetched?